### DISCUSSION TITLE
Error when updating to Alpine 3.12.1

### DISCUSSION THREAD

**Discussion by marcreichel** [alpinejs/alpine#3573] [7. Bugs (archived)] [answered]
I run into the following error when updating to Alpine 3.12.1. When downgrading to Alpine 3.12.0 everything works as expected.

<img width="384" alt="image" src="https://github.com/alpinejs/alpine/assets/7645035/330816bf-3e5a-40bd-9176-3a046a1a7902">

<img width="295" alt="image" src="https://github.com/alpinejs/alpine/assets/7645035/e0c5009b-17b8-4a5e-9f1d-fc49e9cf12ed">

**Comment by ekwoka** [comment]
Can you provide any context to what is happening when you get this error?

it shouldn't be possible

**Comment by marcreichel** [comment]
@ekwoka Yeah, sure. Some parts in my app stopped working completely. Sadly the project I'm working on is closed source, so I cannot provide any reproduction repository. I'm not quite sure yet what is causing the issue. Simple things like `x-text` bindings are still working normally. But it definitely has something to do with some changes which happened between versions 3.12.0 and 3.12.1 because without changing anything else but switching between those two versions everything is working with version 3.12.0, but things break down with version 3.12.1.

**Reply by SimoTod** [reply]
Are you building from npm? I suspect it may happen if the version of vue/reactivity is not the one expected by Alpine. Alpine uses an old version and I think there were breaking changes in one of the minor release of vue/reactivity

**Reply by marcreichel** [reply]
Yes, we're building from NPM, and we have installed Vue version 2.7.14 (which we're going to drop in a few weeks when we have migrated everything from Vue to Alpine). So if this really is the problem here I may pin Alpine to version 3.12.0 for now and unpin it as soon as we're dropping Vue.

**Reply by SimoTod** [reply]
@ekwoka may know better than mee but I believe that whilst the previous version of Alpine came packed with all the dependencies, the latest one doesn't and uses whatever version is available in your node_modules folder. I'm pretty much sure that Alpine is only compatible with `@vue/rectivity 3.1.*` and breaks with that error if you use 3.2 or 3.3 (which, I assume, are the ones imported by Vue).

**Reply by ekwoka** [reply]
Yes, Alpine no longer bundles deps, and it needs the version indicates, which I think might be indicated improperly in the package json preventing npm from complaining.

If you use pnpm, I believe it should allow Alpine to use the one it indicates even if you are using a different version elsewhere.

I'll check all the deps and see about at least ensuring the package.json is pinned so that npm will wave a flag if you have mixed up deps.

Might even be worth finding out what breaking changes happened in Vue reactivity.

**Comment by tonysm** [comment]
I started getting an error on 3.12.1 which I think it's related to this. I'm using import maps (no build tooling):

![image](https://github.com/alpinejs/alpine/assets/1178621/04136f8c-5c3e-494b-9050-1d1805049c88)

This doesn't happen in 3.12.0.

**Reply by ekwoka** [reply]
Where are you importing it from?

That file indicated is not a file in Alpine

**Reply by tonysm** [reply]
I thought it wasn't related, but that error seems to be from `@vue/shared` which is now being pulled from Alpine's ESM module. From version 3.12.0, it doesn't pull that. This Vue package seems to use `process.env`, which relies on having a bundler or something of the like to process it before it's served to the browser (it's not browser-ready). I think it's because of [this change](https://github.com/alpinejs/alpine/pull/3459).

Here's my importmap file after pinning Alpine's latest version:

```php
<?php

use Tonysm\ImportmapLaravel\Facades\Importmap;

Importmap::pinAllFrom("resources/js", to: "js/", preload: true);

// ...

Importmap::pin("alpinejs", to: "https://ga.jspm.io/npm:alpinejs@3.12.1/dist/module.esm.js");
Importmap::pin("@vue/reactivity", to: "https://ga.jspm.io/npm:@vue/reactivity@3.1.4/dist/reactivity.esm-bundler.js");
Importmap::pin("@vue/shared", to: "https://ga.jspm.io/npm:@vue/shared@3.1.4/dist/shared.esm-bundler.js");
```

And when I specifically pull version 3.12.0 (only pulls Alpine:

```php
<?php

use Tonysm\ImportmapLaravel\Facades\Importmap;

Importmap::pinAllFrom("resources/js", to: "js/", preload: true);

// ...

Importmap::pin("alpinejs", to: "https://ga.jspm.io/npm:alpinejs@3.12.0/dist/module.esm.js");
```

Btw, I'm using the [Importmap Laravel](https://github.com/tonysm/importmap-laravel) package I built, but it's based on Rails' importmap-rails gem, so they might be getting that as well when using Alpine with import maps.

Both implementations make use of [JSPM's API](https://jspm.org/), so you can actually try it with cURL:

```curl
curl -Ss -d '{"install": ["alpinejs@3.12.1"], "flattenScope": true, "env": ["browser", "module", "production"], "provider": "jspm"}' -H "Content-Type: application/json" -XPOST https://api.jspm.io/generate | jq
{
  "staticDeps": [
    "https://ga.jspm.io/npm:@vue/reactivity@3.1.4/dist/reactivity.esm-bundler.js",
    "https://ga.jspm.io/npm:@vue/shared@3.1.4/dist/shared.esm-bundler.js",
    "https://ga.jspm.io/npm:alpinejs@3.12.1/dist/module.esm.js"
  ],
  "dynamicDeps": [],
  "map": {
    "imports": {
      "alpinejs": "https://ga.jspm.io/npm:alpinejs@3.12.1/dist/module.esm.js",
      "@vue/reactivity": "https://ga.jspm.io/npm:@vue/reactivity@3.1.4/dist/reactivity.esm-bundler.js",
      "@vue/shared": "https://ga.jspm.io/npm:@vue/shared@3.1.4/dist/shared.esm-bundler.js"
    }
  }
}
```

And previous version:

```curl
curl -Ss -d '{"install": ["alpinejs@3.12.0"], "flattenScope": true, "env": ["browser", "module", "production"], "provider": "jspm"}' -H "Content-Type: application/json" -XPOST https://api.jspm.io/generate | jq
{
  "staticDeps": [
    "https://ga.jspm.io/npm:alpinejs@3.12.0/dist/module.esm.js"
  ],
  "dynamicDeps": [],
  "map": {
    "imports": {
      "alpinejs": "https://ga.jspm.io/npm:alpinejs@3.12.0/dist/module.esm.js"
    }
  }
}
```

**Reply by ekwoka** [reply]
Those import maps are strange, and it does seem you're targeting the bundler version of Vue reactivity.

You can use `.esm-browser.js` instead.

Or use importer urls that do processing for you.

Like `esm.sh`


In this case it seems it's upstream with Vue not having all the appropriate package.json fields, or at least not in that version.

If we can solve the blocking issues for updating, this may also solve itself.

**Reply by tonysm** [reply]
Yeah, it's weird. Not sure what the "real fix" would look like here. I found [this issue in JSPM's repository](https://github.com/jspm/project/issues/192#issuecomment-1132132660) and looks like the `@vue/reactivity` package doesn't export the browser module by default. We can workaround that by pinning a the CJS version like this:

```php
<?php

use Tonysm\ImportmapLaravel\Facades\Importmap;

Importmap::pinAllFrom("resources/js", to: "js/", preload: true);

// ...

Importmap::pin("alpinejs", to: "https://ga.jspm.io/npm:alpinejs@3.12.1/dist/module.cjs.js");
Importmap::pin("@vue/reactivity", to: "https://ga.jspm.io/npm:@vue/reactivity@3.1.4/indexjs");
Importmap::pin("@vue/shared", to: "https://ga.jspm.io/npm:@vue/shared@3.1.4/indexjs");
```

But since this is using the CJS module, I have to tweak my Alpine setup:

```js
import AlpineCJS from 'alpinejs';

window.Alpine = AlpineCJS.default;

Alpine.start();
// ...
```

I figured this out by playing with the [JSPM generator](https://generator.jspm.io/#U2NiYGBkDM0rySzJSU1hSMwpyMxLzSp2MNYzNNIzBAD+yFbRHgA) (for anyone else facing this, I just toggled the "module" env conditions at the bottom-left corner).

Not sure how to properly solve this this. If anyone else has a clue, I'd love to know about it.

Do we really need to define that as an external dependency?

**Reply by ekwoka** [reply]
If you just change where you map the import to `esm-bundler` to `esm-browser` it will work

**Reply by tonysm** [reply]
I'd have to switch to another CDN for that to work. JSPM returns a 404.

**Reply by ekwoka** [reply]
Looks like you can make a PR to https://github.com/jspm/overrides

and it can let you change the `module` entry points on @vue/reactivity to the browser one instead of the bundler one. strange they don't allow overrides in the url params like esm.sh does.

It's a very strange thing that Vue did, but for a mostly internal package I guess it doesn't matter to them that much.

But I guess add this to the list of reasons not to use general CDNs for anything you care about.

**Reply by tonysm** [reply]
Done: https://github.com/jspm/overrides/pull/16

**Comment by nikitaourazbaev** [comment]
I’m seeing the same `process is not defined` error as above building from npm using rollup. 

I’ve made a minimal example repo here:

https://github.com/nikitaourazbaev/alpine-3121-rollup-minimal-example

![CleanShot 2023-05-26 at 14 53 58@2x](https://github.com/alpinejs/alpine/assets/41582891/3ca55bd7-04f2-47b0-802e-b8408dd9b4b8)
![CleanShot 2023-05-26 at 14 54 05@2x](https://github.com/alpinejs/alpine/assets/41582891/50378740-d8c5-44fb-9b06-f8d8666eada8)

Downgrading to 3.12.0 fixes this but I’m not sure how to fix it / work around it going forward. The versions of `@vue/reactivity` and `@vue/shared` installed in `node_modules` are `3.1.5`. There are no other versions of either package.

**Reply by SimoTod** [reply]
it's technically a VueJS issue but you should be able to configure rollup in this way: https://github.com/rollup/rollup/issues/805

**Reply by nikitaourazbaev** [reply]
Thank you, will give that a try. *Update*: That worked!

**Comment by joshhanley** [comment]
@SimoTod and @ekwoka I don’t really follow why this is causing issues in people’s apps, but it seems to be causing problems for people. Do either of you think we should revert this change?

**Reply by ekwoka** [reply]
@joshhanley the reason it's causing issues is that Vue reactivity points to a server version of their code, with a browser version on the side. Most bundlers cns understand what this means and handle it properly, but some cdn sites don't follow modern rules or care and try to put the server code into the browser. This doesn't impact anyone actually bundling their scripts or using modern tooling.

I was about to make a PR for it, but I believe just explicitely importing the browser version instead of the generic entrypoint would solve it everywhere.

So just `@vue/reactivity` to `@vue/reactivity/dist/reactivity.browser-esm.js` or whatever it is.

I can get that together today :)


Of note: one of my clients had this change be breaking for them, but it was because of a misconfiguration in one buildstep. The repo was a monorepo of sorts and all of the contents were configured properly except one. And this change exposed that issue

**Reply by joshhanley** [reply]
@ekwoka thanks for the explanation 🙂 

Ok if you can that'd be great. Whichever you think is better, either the change you suggested or revert. When it's ready, I Will see if Caleb can have a look. 

Edit: just saw your PR, will have a look at it 🙂

**Comment by calebporzio** [comment, marked as answer]
I just reverted the "exclude external dependancies" change (#3459) which should have fixed it. New version tagged: v3.12.2

**Reply by marcreichel** [reply]
@calebporzio Thank you very much! I just updated to `3.12.2` and I can confirm that everything is working as normal now.

**Reply by calebporzio** [reply]
Great, thanks for confirming!
